presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/MultiJoinNode.java

```diff
@@ -0,0 +1,120 @@
+import com.facebook.presto.sql.planner.iterative.Lookup;
+import com.facebook.presto.sql.planner.plan.JoinNode;
+import com.facebook.presto.sql.planner.plan.PlanNode;
+import com.facebook.presto.sql.tree.Expression;
+import com.google.common.collect.ImmutableList;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import static com.facebook.presto.sql.ExpressionUtils.and;
+import static com.facebook.presto.sql.planner.DeterminismEvaluator.isDeterministic;
+import static com.facebook.presto.sql.planner.plan.JoinNode.Type.INNER;
+import static com.facebook.presto.sql.tree.BooleanLiteral.TRUE_LITERAL;
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkState;
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static java.util.Objects.requireNonNull;
+
+/**
+ * This class represents a set of inner joins that can be executed in any order.
+ */
+class MultiJoinNode
+{
+    private final List<PlanNode> sources;
+    private final Expression filter;
+    private final List<Symbol> outputSymbols;
+
+    public MultiJoinNode(List<PlanNode> sources, Expression filter, List<Symbol> outputSymbols)
+    {
+        this.sources = ImmutableList.copyOf(requireNonNull(sources, "sources is null"));
+        this.filter = requireNonNull(filter, "filter is null");
+        this.outputSymbols = ImmutableList.copyOf(requireNonNull(outputSymbols, "outputSymbols is null"));
+
+        List<Symbol> inputSymbols = sources.stream().flatMap(source -> source.getOutputSymbols().stream()).collect(toImmutableList());
+        checkArgument(inputSymbols.containsAll(outputSymbols), "inputs do not contain all output symbols");
+    }
+
+    public Expression getFilter()
+    {
+        return filter;
+    }
+
+    public List<PlanNode> getSources()
+    {
+        return sources;
+    }
+
+    public List<Symbol> getOutputSymbols()
+    {
+        return outputSymbols;
+    }
+
+    static MultiJoinNode toMultiJoinNode(JoinNode joinNode, Lookup lookup, int joinLimit)
+    {
+        return new MultiJoinNodeBuilder(joinNode, lookup, joinLimit).toMultiJoinNode();
+    }
+
+    private static class MultiJoinNodeBuilder
+    {
+        private final List<PlanNode> sources = new ArrayList<>();
+        private final List<Expression> filters = new ArrayList<>();
+        private final List<Symbol> outputSymbols;
+        private final Lookup lookup;
+
+        MultiJoinNodeBuilder(JoinNode node, Lookup lookup, int joinLimit)
+        {
+            requireNonNull(node, "node is null");
+            checkState(node.getType() == INNER, "join type must be INNER");
+            this.outputSymbols = node.getOutputSymbols();
+            this.lookup = requireNonNull(lookup, "lookup is null");
+            flattenNode(node, joinLimit);
+        }
+
+        private void flattenNode(PlanNode node, int limit)
+        {
+            PlanNode resolved = lookup.resolve(node);
```

Comment:
There isn't viable alternative yet. 
(Technically, i can use `resolveGroup` and handle the returned `Stream` to avoid deprecation warning. Would it be anything better?)

Reply:
@findepi My preference would be to either drop `@deprecated` from this method or not use it in the new code. It is confusing to mark something deprecated when there is no alternative.

Reply:
Well -- this method shouldn't generally be used, pattern matching & captures is the recommended way (although I think I  wouldn't deprecate it just yet). However, this case is special, because this rule is recursive and pattern matching doesn't support recursive patterns.
Technically, we could have a separate rule that incrementally builds a `MultiJoinNode` and only then run `ReorderJoins` (in a separate `IterativeOptimizer` instance), but I don't think we should explore this path now.  We have already considered this option but decided not to purse it.

Note also the reasons why pattern matching is recommended over `Lookup#resolve`. This is because we anticipate memo-based optimizer and want the rule code to be portable/adaptable to it easily. However, `ReorderJoins` is a bit special rule and I don't think it would be directly portable to memo-based optimizer, as it could very easily cause an explosion in computational cost.

Reply:
@findepi Piotr, thanks for explaining. Can I read about memo-based optimizer somewhere?

Reply:
@mbasmanova, here's a document where I started capturing some ideas when I began working on this a couple of years ago: https://github.com/prestodb/presto/wiki/New-Optimizer

There some references to papers that could help give some context.